core/log/profiler_hook.cpp

```diff
@@ -0,0 +1,302 @@
+        this->begin_hook_(ss.str().c_str(), profile_event_category::linop);
+    }
+
+    void on_linop_advanced_apply_completed(const LinOp* A, const LinOp* alpha,
+                                           const LinOp* b, const LinOp* beta,
+                                           const LinOp* x) const override
+    {
+        std::stringstream ss;
+        ss << "advanced_apply(" << name_demangling::get_dynamic_type(*A) << ")";
+        this->end_hook_(ss.str().c_str(), profile_event_category::linop);
+    }
+
+    /* LinOpFactory events */
+    void on_linop_factory_generate_started(const LinOpFactory* factory,
+                                           const LinOp* input) const override
+    {
+        std::stringstream ss;
+        ss << "generate(" << name_demangling::get_dynamic_type(*factory) << ")";
+        this->begin_hook_(ss.str().c_str(), profile_event_category::factory);
+    }
+
+    void on_linop_factory_generate_completed(const LinOpFactory* factory,
+                                             const LinOp* input,
+                                             const LinOp* output) const override
+    {
+        std::stringstream ss;
+        ss << "generate(" << name_demangling::get_dynamic_type(*factory) << ")";
+        this->end_hook_(ss.str().c_str(), profile_event_category::factory);
+    }
+
+    /* Criterion events */
+    void on_criterion_check_started(const stop::Criterion* criterion,
+                                    const size_type& num_iterations,
+                                    const LinOp* residual,
+                                    const LinOp* residual_norm,
+                                    const LinOp* solution,
+                                    const uint8& stopping_id,
+                                    const bool& set_finalized) const override
+    {
+        std::stringstream ss;
+        ss << "check(" << name_demangling::get_dynamic_type(*criterion) << ")";
+        this->begin_hook_(ss.str().c_str(), profile_event_category::criterion);
+    }
+
+    void on_criterion_check_completed(
+        const stop::Criterion* criterion, const size_type& num_iterations,
+        const LinOp* residual, const LinOp* residual_norm,
+        const LinOp* solutino, const uint8& stopping_id,
+        const bool& set_finalized, const array<stopping_status>* status,
+        const bool& one_changed, const bool& all_converged) const override
+    {
+        std::stringstream ss;
+        ss << "check(" << name_demangling::get_dynamic_type(*criterion) << ")";
+        this->end_hook_(ss.str().c_str(), profile_event_category::criterion);
+    }
+
+    ProfilerHook(hook_function begin, hook_function end)
+        : begin_hook_{begin}, end_hook_{end}
+    {}
+
+private:
+    hook_function begin_hook_;
+    hook_function end_hook_;
+};
+
+
+std::mutex profiler_hook_mutex{};
+
+
+std::shared_ptr<Logger> get_tau_hook(bool initialize)
```

Comment:
All of the loggers have different requirements and feature sets (global init/finalize, range attributes/colors, ...), so to put all behind the same interface would mean adding all of these options to all of them (most of them having no effect) or restricting otherwise useful features (colored ranges for example)